graphql/batch_executor.go

```diff
@@ -335,27 +381,23 @@ func resolveObjectBatch(ctx context.Context, sources []interface{}, typ *Object,
 		}
 
 		field := typ.Fields[selection.Name]
-		if field.Expensive {
-			// Expensive fields should be executed as multiple "Units".  The scheduler
-			// controls how the units are executed
-			for idx, source := range nonNilSources {
-				workUnits = append(workUnits, &WorkUnit{
-					ctx:          ctx,
-					field:        field,
-					sources:      []interface{}{source},
-					destinations: []*outputNode{destForSelection[idx]},
-					selection:    selection,
-				})
-			}
-			continue
-		}
 		unit := &WorkUnit{
 			ctx:          ctx,
 			field:        field,
 			sources:      nonNilSources,
 			destinations: destForSelection,
 			selection:    selection,
 		}
+		if shouldUseBatch(selection, field) {
```

Comment:
```go
switch {
case shouldUseBatch(selection, field):
			workUnits = append(workUnits, unit)
case filed.Expensive:
			// Expensive fields should be executed as multiple "Units".  The scheduler
			// controls how the units are executed
			workUnits = append(workUnits, splitWorkUnit(unit)...)
case field.External:
			// External non-Expensive fields should be fast (so we can run them at the
			// same time), but, since they are still external functions we don't want
			// to run them where they could potentially block.
			// So we create an work unit with all the fields to execute
			// asynchronously.
			workUnits = append(workUnits, unit)
default:
		// If the fields are not expensive or external the work time should be
		// bounded, so we can resolve them immediately.
		workUnits = append(
			workUnits,
			executeWorkUnit(unit)...,
		)
}
```